File: archive/quoting-go/bfs.go
package main

import (
	"math/big"

	"github.com/ava-labs/libevm/common"
)

type BfsResult struct {
	AmountOut    *big.Int
	Route        []RouteStep
	GasUsed      uint64
	EvmCalls     int
	RequoteCalls int
}

type bestCandidate struct {
	amountOut *big.Int
	gasTotal  uint64
	route     []RouteStep
}

type backtrackEntry struct {
	pool      common.Address
	poolType  uint8
	fromToken common.Address
	amountIn  *big.Int
	amountOut *big.Int
	extraData ExtraData
}

func bfsRoute(
	evm *LocalEVM,
	quoterCode []byte,
	tokenOverrides map[common.Address]*TokenOverride,
	poolsByToken map[common.Address][]PoolEdge,
	tokenIn, tokenOut common.Address,
	inputAmount *big.Int,
	inputAvax *big.Int,
	maxHops int,
	blockInfo *BlockInfo,
	gasLimit uint64,
	gasAware bool,
	requote bool,
) *BfsResult {
	gasAware = gasAware && inputAvax.Sign() > 0
	basefee := blockInfo.Basefee

	// frontier: token -> (amount, cumulative gas)
	type frontierEntry struct {
		amount *big.Int
		gas    uint64
	}
	frontier := map[common.Address]frontierEntry{
		tokenIn: {amount: new(big.Int).Set(inputAmount), gas: 0},
	}

	totalEvmCalls := 0
	totalRequoteCalls := 0

	backtrack := make([]map[common.Address]*backtrackEntry, 0, maxHops)
	bestAtLayer := make([]*bestCandidate, maxHops)

	reachable := buildReachability(poolsByToken, tokenOut, maxHops)

	for layer := 0; layer < maxHops; layer++ {
		isLastLayer := layer == maxHops-1
		remaining := maxHops - layer - 1

		// Build jobs
		var jobs []QuoteJob
		for token, fe := range frontier {
			edges, ok := poolsByToken[token]
			if !ok {
				continue
			}
			for _, edge := range edges {
				// Same-pool reversal pruning
				if layer > 0 {
					if bt, ok := backtrack[layer-1][token]; ok {
						if edge.TokenOut == bt.fromToken && edge.Pool == bt.pool {
							continue
						}
					}
				}
				if edge.TokenOut == tokenIn && tokenIn != tokenOut {
					continue
				}
				if isLastLayer && edge.TokenOut != tokenOut {
					continue
				}
				if !isLastLayer && remaining > 0 {
					if remaining-1 < len(reachable) {
						if _, ok := reachable[remaining-1][edge.TokenOut]; !ok {
							continue
						}
					}
				}
				jobs = append(jobs, QuoteJob{
					Pool:      edge.Pool,
					PoolType:  edge.PoolType,
					TokenIn:   edge.TokenIn,
					TokenOut:  edge.TokenOut,
					Amount:    fe.amount,
					ExtraData: edge.ExtraData,
				})
			}
		}

		if len(jobs) == 0 {
			backtrack = append(backtrack, make(map[common.Address]*backtrackEntry))
			continue
		}

		nextFrontier := make(map[common.Address]frontierEntry)
		layerBacktrack := make(map[common.Address]*backtrackEntry)

		type terminalCandidate struct {
			estimatedAmount *big.Int
			estGas          uint64
			pool            common.Address
			poolType        uint8
			fromToken       common.Address
			extraData       ExtraData
		}
		var tokenOutCandidates []terminalCandidate

		// Quote all edges
		results, numCalls := quoteAll(evm, quoterCode, tokenOverrides, jobs, blockInfo, gasLimit)
		totalEvmCalls += numCalls

		for _, result := range results {
			if result.AmountOut.Sign() == 0 {
				continue
			}
			fromGas := uint64(0)
			if fe, ok := frontier[result.TokenIn]; ok {
				fromGas = fe.gas
			}
			pathGas := fromGas + result.GasUsed

			if result.TokenOut == tokenOut {
				if layer == 0 {
					// Layer 0: single-hop quotes are exact, score directly
					if isBetter(bestAtLayer[layer], result.AmountOut, pathGas, gasAware, inputAvax, basefee) {
						bestAtLayer[layer] = &bestCandidate{
							amountOut: result.AmountOut, gasTotal: pathGas,
							route: []RouteStep{{
								Pool: result.Pool, PoolType: result.PoolType,
								TokenIn: result.TokenIn, TokenOut: tokenOut,
								AmountIn: result.Amount, AmountOut: result.AmountOut,
								ExtraData: result.ExtraData,
							}},
						}
					}
				} else {
					// Layer >= 1: queue for full-path atomic quoting below
					tokenOutCandidates = append(tokenOutCandidates, terminalCandidate{
						estimatedAmount: result.AmountOut, estGas: pathGas,
						pool: result.Pool, poolType: result.PoolType,
						fromToken: result.TokenIn, extraData: result.ExtraData,
					})
				}
			} else {
				existing, exists := nextFrontier[result.TokenOut]
				better := !exists
				if exists {
					if gasAware {
						better = isBetterNet(result.AmountOut, pathGas, existing.amount, existing.gas, inputAvax, basefee)
					} else {
						better = result.AmountOut.Cmp(existing.amount) > 0
					}
				}
				if better {
					nextFrontier[result.TokenOut] = frontierEntry{amount: result.AmountOut, gas: pathGas}
					layerBacktrack[result.TokenOut] = &backtrackEntry{
						pool: result.Pool, poolType: result.PoolType,
						fromToken: result.TokenIn, amountIn: result.Amount,
						amountOut: result.AmountOut, extraData: result.ExtraData,
					}
				}
			}
		}

		backtrack = append(backtrack, layerBacktrack)

		// FULL-PATH ATOMIC QUOTING (layer >= 1)
		//
		// For multi-hop routes, we quote the entire accumulated path atomically
		// via executeRoute() instead of relying on single-hop estimates.
		// Single-hop swap above is used only as a FILTER to identify
		// which edges are worth exploring. The actual amounts come from here.
		//
		// Why: executing hops atomically means hop1 mutates pool state that
		// hop2 reads. Single-hop estimates ignore this, causing ~0.01% drift.
		//
		// TODO: This is a correctness-first approach. The proper fix is to
		// propagate state diffs through the route during BFS expansion,
		// avoiding the need to re-execute the full prefix for every new edge.
		// Current approach roughly doubles execution time.
		if layer > 0 {
			// Build all route jobs: both frontier and terminal candidates
			// are executed in a single parallel batch (matching Rust's approach).
			var allJobs []RouteJob
			var frontierTokens []common.Address
			var terminalInfos []terminalCandidate

			// Frontier routes
			for token := range nextFrontier {
				route := reconstructPath(backtrack, token, tokenIn, inputAmount)
				if route == nil {
					continue
				}
				frontierTokens = append(frontierTokens, token)
				allJobs = append(allJobs, RouteJob{Route: route})
			}

			terminalStart := len(allJobs)

			// Terminal routes
			for _, cand := range tokenOutCandidates {
				var prefix []RouteStep
				if cand.fromToken == tokenIn {
					prefix = nil
				} else {
					prefix = reconstructPath(backtrack, cand.fromToken, tokenIn, inputAmount)
					if prefix == nil {
						continue
					}
				}
				route := append(prefix, RouteStep{
					Pool: cand.pool, PoolType: cand.poolType,
					TokenIn: cand.fromToken, TokenOut: tokenOut,
					AmountIn: big.NewInt(0), AmountOut: big.NewInt(0),
					ExtraData: cand.extraData,
				})
				if len(route) > 0 {
					route[0].AmountIn = inputAmount
				}
				terminalInfos = append(terminalInfos, cand)
				allJobs = append(allJobs, RouteJob{Route: route})
			}

			if len(allJobs) > 0 {
				// Execute all routes in parallel
				results := executeRoutesParallel(evm, quoterCode, tokenOverrides, allJobs, inputAmount, blockInfo, gasLimit)
				totalRequoteCalls += len(results)

				// Apply frontier results
				for i := 0; i < terminalStart; i++ {
					token := frontierTokens[i]
					if results[i].AmountOut.Sign() > 0 {
						nextFrontier[token] = frontierEntry{amount: results[i].AmountOut, gas: results[i].GasUsed}
					} else {
						delete(nextFrontier, token)
					}
				}

				// Apply terminal results
				for i := terminalStart; i < len(results); i++ {
					if results[i].AmountOut.Sign() == 0 {
						continue
					}
					if isBetter(bestAtLayer[layer], results[i].AmountOut, results[i].GasUsed, gasAware, inputAvax, basefee) {
						bestAtLayer[layer] = &bestCandidate{
							amountOut: results[i].AmountOut, gasTotal: results[i].GasUsed,
							route: allJobs[i].Route,
						}
					}
				}
			}
		}

		// Update frontier
		frontier = make(map[common.Address]frontierEntry)
		for token, fe := range nextFrontier {
			if fe.amount.Sign() > 0 {
				frontier[token] = fe
			}
		}
	}

	// Pick best across all layers
	var bestLayer int = -1
	bestAmount := big.NewInt(0)
	var bestGas uint64

	for layer, cand := range bestAtLayer {
		if cand == nil {
			continue
		}
		better := bestLayer == -1
		if !better {
			if gasAware {
				better = isBetterNet(cand.amountOut, cand.gasTotal, bestAmount, bestGas, inputAvax, basefee)
			} else {
				better = cand.amountOut.Cmp(bestAmount) > 0
			}
		}
		if better {
			bestLayer = layer
			bestAmount = cand.amountOut
			bestGas = cand.gasTotal
		}
	}

	// Use the stored route directly — no reconstruction needed
	var route []RouteStep
	if bestLayer >= 0 {
		route = bestAtLayer[bestLayer].route
	}

	return &BfsResult{
		AmountOut:    bestAmount,
		Route:        route,
		GasUsed:      bestGas,
		EvmCalls:     totalEvmCalls,
		RequoteCalls: totalRequoteCalls,
	}
}

// ---- Helpers ----

func isBetter(current *bestCandidate, newAmount *big.Int, newGas uint64, gasAware bool, inputAvax *big.Int, basefee uint64) bool {
	if current == nil {
		return true
	}
	if gasAware {
		return isBetterNet(newAmount, newGas, current.amountOut, current.gasTotal, inputAvax, basefee)
	}
	return newAmount.Cmp(current.amountOut) > 0
}

func buildReachability(poolsByToken map[common.Address][]PoolEdge, tokenOut common.Address, maxHops int) []map[common.Address]bool {
	// Build reverse map: tokenOut -> set of tokenIn
	reverse := make(map[common.Address]map[common.Address]bool)
	for _, edges := range poolsByToken {
		for _, e := range edges {
			if reverse[e.TokenOut] == nil {
				reverse[e.TokenOut] = make(map[common.Address]bool)
			}
			reverse[e.TokenOut][e.TokenIn] = true
		}
	}

	layers := make([]map[common.Address]bool, 0, maxHops)
	currentSet := map[common.Address]bool{tokenOut: true}
	if sources, ok := reverse[tokenOut]; ok {
		for t := range sources {
			currentSet[t] = true
		}
	}
	layers = append(layers, copyBoolSet(currentSet))

	for i := 1; i < maxHops; i++ {
		nextSet := copyBoolSet(currentSet)
		for t := range currentSet {
			if sources, ok := reverse[t]; ok {
				for s := range sources {
					nextSet[s] = true
				}
			}
		}
		currentSet = nextSet
		layers = append(layers, copyBoolSet(currentSet))
	}

	return layers
}

func copyBoolSet(m map[common.Address]bool) map[common.Address]bool {
	r := make(map[common.Address]bool, len(m))
	for k := range m {
		r[k] = true
	}
	return r
}

func reconstructPath(backtrack []map[common.Address]*backtrackEntry, toToken, tokenIn common.Address, inputAmount *big.Int) []RouteStep {
	if len(backtrack) == 0 {
		return nil
	}
	lastLayer := len(backtrack) - 1
	entry, ok := backtrack[lastLayer][toToken]
	if !ok {
		return nil
	}

	var steps []RouteStep
	steps = append(steps, RouteStep{
		Pool: entry.pool, PoolType: entry.poolType,
		TokenIn: entry.fromToken, TokenOut: toToken,
		AmountIn: entry.amountIn, AmountOut: entry.amountOut,
		ExtraData: entry.extraData,
	})

	current := entry.fromToken
	for l := lastLayer - 1; l >= 0; l-- {
		if current == tokenIn {
			break
		}
		e, ok := backtrack[l][current]
		if !ok {
			return nil
		}
		steps = append(steps, RouteStep{
			Pool: e.pool, PoolType: e.poolType,
			TokenIn: e.fromToken, TokenOut: current,
			AmountIn: e.amountIn, AmountOut: e.amountOut,
			ExtraData: e.extraData,
		})
		current = e.fromToken
	}

	if current != tokenIn {
		return nil
	}

	// Reverse
	for i, j := 0, len(steps)-1; i < j; i, j = i+1, j-1 {
		steps[i], steps[j] = steps[j], steps[i]
	}
	if len(steps) > 0 {
		steps[0].AmountIn = inputAmount
	}
	return steps
}

File: archive/pathfinders/simple-bfs/src/bfs.rs
use alloy_primitives::{Address, B256, U256};
use std::collections::{HashMap, HashSet};
use std::sync::Arc;

use crate::overrides;
use crate::quoter::{
    self, BlockInfo, ExtraData, PoolsByToken, QuoteJob, RouteStep,
};
use crate::state::{RpcFetcher, StateSnapshot};

pub struct BfsResult {
    pub amount_out: U256,
    pub route: Vec<RouteStep>,
    pub gas_used: u64,
    pub evm_calls: usize,
    pub requote_calls: usize,
    pub fetched_storage: Vec<(Address, B256, U256)>,
    pub fetched_accounts: Vec<(Address, U256, B256, revm::bytecode::Bytecode)>,
}

pub fn bfs_route(
    rayon_pool: &rayon::ThreadPool,
    snapshot: &Arc<StateSnapshot>,
    rpc_fetcher: &Arc<RpcFetcher>,
    contract_code: &[u8],
    token_overrides: &HashMap<Address, overrides::TokenOverride>,
    pools_by_token: &PoolsByToken,
    token_in: Address,
    token_out: Address,
    input_amount: U256,
    input_avax: U256,
    max_hops: usize,
    block_info: &BlockInfo,
    gas_limit: u64,
    gas_aware: bool,
    requote: bool,
) -> BfsResult {
    let gas_aware = gas_aware && !input_avax.is_zero();
    let basefee = block_info.basefee;

    let mut frontier: HashMap<Address, (U256, u64)> = HashMap::new();
    frontier.insert(token_in, (input_amount, 0u64));

    let mut total_evm_calls = 0usize;
    let mut total_requote_calls = 0usize;
    let mut all_fetched_storage = Vec::new();
    let mut all_fetched_accounts = Vec::new();

    let mut backtrack: Vec<HashMap<Address, (Address, u8, Address, U256, U256, ExtraData)>> =
        Vec::with_capacity(max_hops);

    // (amount_out, gas_total, full_route)
    let mut best_at_layer: Vec<Option<(U256, u64, Vec<RouteStep>)>> =
        vec![None; max_hops];

    let reachable = build_reachability(pools_by_token, token_out, max_hops);

    for layer in 0..max_hops {
        let is_last_layer = layer == max_hops - 1;
        let remaining = max_hops - layer - 1;

        // Build jobs
        let mut jobs: Vec<QuoteJob> = Vec::new();
        for (token, (amount, _)) in &frontier {
            if let Some(edges) = pools_by_token.get(token) {
                for edge in edges {
                    // Same-pool reversal pruning
                    if layer > 0 {
                        if let Some((parent_pool, _, parent_token, _, _, _)) =
                            backtrack[layer - 1].get(token)
                        {
                            if edge.token_out == *parent_token && edge.pool == *parent_pool {
                                continue;
                            }
                        }
                    }
                    if edge.token_out == token_in && token_in != token_out {
                        continue;
                    }
                    if is_last_layer && edge.token_out != token_out {
                        continue;
                    }
                    if !is_last_layer && remaining > 0 {
                        if let Some(reachable_set) = reachable.get(remaining - 1) {
                            if !reachable_set.contains(&edge.token_out) {
                                continue;
                            }
                        }
                    }
                    jobs.push(QuoteJob {
                        pool: edge.pool,
                        pool_type: edge.pool_type,
                        token_in: edge.token_in,
                        token_out: edge.token_out,
                        amount: *amount,
                        extra_data: edge.extra_data,
                    });
                }
            }
        }

        if jobs.is_empty() {
            backtrack.push(HashMap::new());
            continue;
        }

        let mut next_frontier: HashMap<Address, (U256, u64)> = HashMap::new();
        let mut layer_backtrack: HashMap<Address, (Address, u8, Address, U256, U256, ExtraData)> =
            HashMap::new();
        let mut token_out_candidates: Vec<(U256, u64, Address, u8, Address, ExtraData)> =
            Vec::new();

        // Quote all edges
        {
            let (results, num_calls, fetched) = quoter::quote_all(
                rayon_pool, snapshot, rpc_fetcher, contract_code,
                token_overrides, &jobs, block_info, gas_limit,
            );
            total_evm_calls += num_calls;
            all_fetched_storage.extend(fetched.storage);
            all_fetched_accounts.extend(fetched.accounts);

            for (_, result) in &results {
                if result.amount_out.is_zero() {
                    continue;
                }
                let from_gas = frontier.get(&result.token_in).map_or(0u64, |&(_, g)| g);
                let path_gas = from_gas.saturating_add(result.gas_used);

                if result.token_out == token_out {
                    if layer == 0 {
                        // Layer 0: single-hop quotes are exact, score directly
                        let better = is_better(
                            &best_at_layer[layer], result.amount_out, path_gas,
                            gas_aware, input_avax, basefee,
                        );
                        if better {
                            best_at_layer[layer] = Some((
                                result.amount_out, path_gas,
                                vec![RouteStep {
                                    pool: result.pool, pool_type: result.pool_type,
                                    token_in: result.token_in, token_out,
                                    amount_in: result.amount, amount_out: result.amount_out,
                                    extra_data: result.extra_data,
                                }],
                            ));
                        }
                    } else {
                        // Layer >= 1: queue for full-path atomic quoting below
                        token_out_candidates.push((
                            result.amount_out, path_gas, result.pool,
                            result.pool_type, result.token_in, result.extra_data,
                        ));
                    }
                } else {
                    let better = match next_frontier.get(&result.token_out) {
                        None => true,
                        Some(&(old_amount, old_gas)) => {
                            if gas_aware {
                                quoter::is_better_net(
                                    result.amount_out, path_gas, old_amount, old_gas,
                                    input_avax, basefee,
                                )
                            } else {
                                result.amount_out > old_amount
                            }
                        }
                    };
                    if better {
                        next_frontier.insert(result.token_out, (result.amount_out, path_gas));
                        layer_backtrack.insert(
                            result.token_out,
                            (result.pool, result.pool_type, result.token_in,
                             result.amount, result.amount_out, result.extra_data),
                        );
                    }
                }
            }
        }

        backtrack.push(layer_backtrack);

        // FULL-PATH ATOMIC QUOTING (layer >= 1)
        //
        // For multi-hop routes, we quote the entire accumulated path atomically
        // via quoteRoute() instead of relying on single-hop estimates.
        // Single-hop quoteMulti above is used only as a FILTER to identify
        // which edges are worth exploring. The actual amounts come from here.
        //
        // Why: executing hops atomically means hop1 mutates pool state that
        // hop2 reads. Single-hop estimates ignore this, causing ~0.01% drift.
        //
        // TODO: This is a correctness-first approach. The proper fix is to
        // propagate state diffs through the route during BFS expansion,
        // avoiding the need to re-execute the full prefix for every new edge.
        // Current approach roughly doubles execution time.
        if layer > 0 {
            // Build all route jobs: both frontier entries and terminal candidates
            // are re-quoted in a single parallel batch.
            let mut all_route_jobs: Vec<(Vec<RouteStep>, U256)> = Vec::new();
            // Track which jobs are frontier vs terminal so we can dispatch results
            let mut frontier_tokens: Vec<Address> = Vec::new();
            let mut terminal_infos: Vec<(Address, u8, Address, ExtraData)> = Vec::new();

            // Frontier routes: reconstruct full path from tokenIn to each intermediate token
            for token in next_frontier.keys() {
                if let Some(route) = reconstruct_path(&backtrack, *token, token_in, input_amount) {
                    frontier_tokens.push(*token);
                    all_route_jobs.push((route, input_amount));
                }
            }

            let terminal_start = all_route_jobs.len();

            // Terminal routes: reconstruct full path from tokenIn through to token_out
            for &(_estimated_amount, _est_gas, pool, pool_type, from_token, extra_data) in
                &token_out_candidates
            {
                let prefix = if from_token == token_in {
                    Some(Vec::new())
                } else {
                    reconstruct_path(&backtrack, from_token, token_in, input_amount)
                };
                let route = match prefix {
                    Some(mut r) => {
                        r.push(RouteStep {
                            pool,
                            pool_type,
                            token_in: from_token,
                            token_out,
                            amount_in: U256::ZERO,
                            amount_out: U256::ZERO,
                            extra_data,
                        });
                        if let Some(first) = r.first_mut() {
                            first.amount_in = input_amount;
                        }
                        r
                    }
                    None => continue,
                };
                terminal_infos.push((pool, pool_type, from_token, extra_data));
                all_route_jobs.push((route, input_amount));
            }

            if !all_route_jobs.is_empty() {
                let (path_results, route_fetched) = quoter::quote_routes_parallel(
                    rayon_pool, snapshot, rpc_fetcher, contract_code,
                    token_overrides, &all_route_jobs, block_info, gas_limit,
                );
                total_requote_calls += path_results.len();
                all_fetched_storage.extend(route_fetched.storage);
                all_fetched_accounts.extend(route_fetched.accounts);

                // Apply frontier results: replace single-hop estimates with atomic amounts
                for (i, &(real_amount, real_gas)) in path_results[..terminal_start].iter().enumerate() {
                    let token = frontier_tokens[i];
                    if !real_amount.is_zero() {
                        next_frontier.insert(token, (real_amount, real_gas));
                    }
                }

                // Apply terminal results: score against best_at_layer using atomic amounts
                for (i, &(real_amount, real_gas)) in path_results[terminal_start..].iter().enumerate() {
                    if real_amount.is_zero() {
                        continue;
                    }
                    let better = is_better(
                        &best_at_layer[layer], real_amount, real_gas,
                        gas_aware, input_avax, basefee,
                    );
                    if better {
                        // Store the full route that was atomically executed
                        let full_route = all_route_jobs[terminal_start + i].0.clone();
                        best_at_layer[layer] = Some((real_amount, real_gas, full_route));
                    }
                }
            }
        }

        frontier = next_frontier
            .into_iter()
            .filter(|(_, (a, _))| *a > U256::ZERO)
            .collect();
    }

    // Pick best across all layers
    let mut best_layer: Option<usize> = None;
    let mut best_amount = U256::ZERO;
    let mut best_gas = 0u64;

    for (layer, candidate) in best_at_layer.iter().enumerate() {
        if let Some((amount_out, gas_total, _)) = candidate {
            let better = match best_layer {
                None => true,
                _ => {
                    if gas_aware {
                        quoter::is_better_net(*amount_out, *gas_total, best_amount, best_gas, input_avax, basefee)
                    } else {
                        *amount_out > best_amount
                    }
                }
            };
            if better {
                best_layer = Some(layer);
                best_amount = *amount_out;
                best_gas = *gas_total;
            }
        }
    }

    // Use the stored route directly — no reconstruction needed
    let route = if let Some(final_layer) = best_layer {
        best_at_layer[final_layer].as_ref().unwrap().2.clone()
    } else {
        Vec::new()
    };

    BfsResult {
        amount_out: best_amount,
        route,
        gas_used: best_gas,
        evm_calls: total_evm_calls,
        requote_calls: total_requote_calls,
        fetched_storage: all_fetched_storage,
        fetched_accounts: all_fetched_accounts,
    }
}

// ---- Helpers ----

fn is_better(
    current: &Option<(U256, u64, Vec<RouteStep>)>,
    new_amount: U256,
    new_gas: u64,
    gas_aware: bool,
    input_avax: U256,
    basefee: u64,
) -> bool {
    match current {
        None => true,
        Some((old_out, old_gas, _)) => {
            if gas_aware {
                quoter::is_better_net(new_amount, new_gas, *old_out, *old_gas, input_avax, basefee)
            } else {
                new_amount > *old_out
            }
        }
    }
}

fn build_reachability(
    pools_by_token: &PoolsByToken,
    token_out: Address,
    max_hops: usize,
) -> Vec<HashSet<Address>> {
    let mut reverse: HashMap<Address, HashSet<Address>> = HashMap::new();
    for edges in pools_by_token.values() {
        for e in edges {
            reverse.entry(e.token_out).or_default().insert(e.token_in);
        }
    }
    let mut layers: Vec<HashSet<Address>> = Vec::with_capacity(max_hops);
    let mut current_set = HashSet::new();
    current_set.insert(token_out);
    if let Some(sources) = reverse.get(&token_out) {
        for &t in sources {
            current_set.insert(t);
        }
    }
    layers.push(current_set.clone());
    for _ in 1..max_hops {
        let mut next_set = current_set.clone();
        for &t in &current_set {
            if let Some(sources) = reverse.get(&t) {
                for &s in sources {
                    next_set.insert(s);
                }
            }
        }
        current_set = next_set;
        layers.push(current_set.clone());
    }
    layers
}

fn reconstruct_path(
    backtrack: &[HashMap<Address, (Address, u8, Address, U256, U256, ExtraData)>],
    to_token: Address,
    token_in: Address,
    input_amount: U256,
) -> Option<Vec<RouteStep>> {
    let last_layer = backtrack.len().checked_sub(1)?;
    let entry = backtrack[last_layer].get(&to_token)?;
    let (pool, pool_type, from_token, amount_in, amount_out, extra_data) = *entry;

    let mut steps = Vec::new();
    steps.push(RouteStep {
        pool,
        pool_type,
        token_in: from_token,
        token_out: to_token,
        amount_in,
        amount_out,
        extra_data,
    });

    let mut current = from_token;
    for layer in (0..last_layer).rev() {
        if current == token_in {
            break;
        }
        let e = backtrack[layer].get(&current)?;
        let (p, pt, bt, ba_in, ba_out, bed) = *e;
        steps.push(RouteStep {
            pool: p,
            pool_type: pt,
            token_in: bt,
            token_out: current,
            amount_in: ba_in,
            amount_out: ba_out,
            extra_data: bed,
        });
        current = bt;
    }

    if current != token_in {
        return None;
    }

    steps.reverse();
    if let Some(first) = steps.first_mut() {
        first.amount_in = input_amount;
    }
    Some(steps)
}
